File: src/domain/planned-vs-executed.js
// ═══════════════════════════════════════════════════════════
// PlannedVsExecuted — avaliação pedagógica do plano
// ═══════════════════════════════════════════════════════════
// Para cada foco do plano (um por PARCIAL), analisa as execuções reais
// daquela parcial e emite veredicto:
//
//   CONSEGUIU         → ação alvo foi observada consistentemente
//   PARCIAL           → só parte das voltas mostrou a ação alvo
//   TENTOU_FALHOU     → tentou a ação mas terminou em erro
//   ERRADO            → não executou a ação planejada (fez outra coisa)
//   SEM_DADOS         → poucas execuções ou sem benchmark

import { ErrorClassifier, ErroTipo } from './error-classifier.js';

export const PvEResultado = Object.freeze({
  CONSEGUIU:      'conseguiu',
  PARCIAL:        'parcial',
  TENTOU_FALHOU:  'tentou-falhou',
  ERRADO:         'errado',
  SEM_DADOS:      'sem-dados',
});

const ACAO_MAP = {
  'atrasar frenagem':   { alinhadas: [ErroTipo.BOA_FRENAGEM, ErroTipo.MANTEVE_LINHA], tentativas: [ErroTipo.FREOU_TARDE_DEMAIS] },
  'antecipar frenagem': { alinhadas: [ErroTipo.MANTEVE_LINHA, ErroTipo.BOA_FRENAGEM], tentativas: [ErroTipo.FREOU_CEDO] },
  'acelerar mais cedo': { alinhadas: [ErroTipo.SAIU_FORTE, ErroTipo.MANTEVE_LINHA], tentativas: [ErroTipo.PERDEU_TRACAO] },
  'saída mais forte':   { alinhadas: [ErroTipo.SAIU_FORTE, ErroTipo.MANTEVE_LINHA], tentativas: [ErroTipo.PERDEU_TRACAO] },
  'entrar mais forte':  { alinhadas: [ErroTipo.BOA_FRENAGEM, ErroTipo.MANTEVE_LINHA], tentativas: [ErroTipo.ENTROU_FORTE_DEMAIS] },
  'manter linha':       { alinhadas: [ErroTipo.MANTEVE_LINHA], tentativas: [] },
  'reduzir tempo':      { alinhadas: [ErroTipo.MANTEVE_LINHA, ErroTipo.BOA_FRENAGEM, ErroTipo.SAIU_FORTE], tentativas: [] },
  'trabalhar execução': { alinhadas: [ErroTipo.MANTEVE_LINHA, ErroTipo.BOA_FRENAGEM, ErroTipo.SAIU_FORTE], tentativas: [] },
  'focar em':           { alinhadas: [ErroTipo.MANTEVE_LINHA, ErroTipo.BOA_FRENAGEM, ErroTipo.SAIU_FORTE], tentativas: [] },
};

function mapAcao(acao) {
  if (!acao) return null;
  const lower = acao.toLowerCase();
  for (const key of Object.keys(ACAO_MAP)) {
    if (lower.includes(key)) return ACAO_MAP[key];
  }
  return null;
}

function _aggregateVelocidades(execs) {
  const pick = (field) => {
    const vals = execs.map(e => e?.[field]).filter(v => v != null && isFinite(v));
    if (vals.length === 0) return null;
    return vals.reduce((a, b) => a + b, 0) / vals.length;
  };
  return {
    velEntradaMedia: pick('velEntrada'),
    velMinimaMedia:  pick('velMinima'),
    velSaidaMedia:   pick('velSaida'),
  };
}

function buildReference(parcialId, benchmark, execsPorParcial = {}) {
  const melhor = benchmark?.melhorPorParcial?.[parcialId]?.tempoMs;
  const vels = _aggregateVelocidades(execsPorParcial[parcialId] || []);
  return { tempoMs: melhor ?? null, ...vels };
}

function buildExecution(parcialId, lap, execsDoLap = []) {
  const execParcial = execsDoLap.find(e => e.parcialId === parcialId);
  return {
    tempoMs: lap.temposPorParcial?.[parcialId] ?? null,
    velEntrada: execParcial?.velEntrada ?? null,
    velMinima:  execParcial?.velMinima  ?? null,
    velSaida:   execParcial?.velSaida   ?? null,
  };
}

export const PlannedVsExecuted = {
  PvEResultado,

  /**
   * Avalia o plano contra as voltas da sessão.
   * @param {object} plan       - PedagogicalPlan (focos[].parcialId)
   * @param {Array}  laps
   * @param {object} benchmark  - { parciais, melhorPorParcial, ... }
   */
  evaluate({ plan, laps, benchmark } = {}) {
    if (!plan || !Array.isArray(plan.focos)) return null;
    const validLaps = (laps || []).filter(l => l.valida);

    const focos = plan.focos.map(foco => {
      const { parcialId, parcialNome, acaoPrincipal } = foco;
      const mapping = mapAcao(acaoPrincipal);

      const execucoes = validLaps.map(l => {
        const execution = buildExecution(parcialId, l);
        const reference = buildReference(parcialId, benchmark);
        let classificacao = null;
        if (reference.tempoMs != null && execution.tempoMs != null) {
          const dTempo = (execution.tempoMs - reference.tempoMs) / reference.tempoMs;
          if (dTempo <= 0.01) classificacao = { principal: ErroTipo.MANTEVE_LINHA };
          else if (dTempo <= 0.03) classificacao = { principal: ErroTipo.BOA_FRENAGEM };
          else classificacao = { principal: null };
        }
        return { lapId: l.id, numero: l.numero, tempoMs: execution.tempoMs, classificacao };
      });

      const classificados = execucoes.filter(e => e?.classificacao?.principal != null);
      const resultado = _julgar(mapping, classificados);
      const proximaAcao = _sugerirProxima(resultado, foco);

      return {
        parcialId,
        parcialNome,
        acaoAlvo: acaoPrincipal,
        justificativa: foco.justificativa,
        execucoes,
        resultado,
        proximaAcao,
      };
    });

    return {
      metaPrincipal: plan.metaPrincipal,
      criterioSucesso: plan.criterioSucesso,
      focos,
    };
  },
};

function _julgar(mapping, classificados) {
  if (classificados.length === 0) return PvEResultado.SEM_DADOS;
  if (!mapping) {
    const ok = classificados.filter(c => c.classificacao.principal === ErroTipo.MANTEVE_LINHA).length;
    const pct = ok / classificados.length;
    if (pct >= 0.6) return PvEResultado.PARCIAL;
    return PvEResultado.SEM_DADOS;
  }
  const n = classificados.length;
  const alinhadas = classificados.filter(c => mapping.alinhadas.includes(c.classificacao.principal)).length;
  const tentativas = classificados.filter(c => mapping.tentativas.includes(c.classificacao.principal)).length;
  const pctAlinhadas = alinhadas / n;
  const pctTentativas = tentativas / n;

  if (pctAlinhadas >= 0.6) return PvEResultado.CONSEGUIU;
  if (pctAlinhadas >= 0.3) return PvEResultado.PARCIAL;
  if (pctTentativas >= 0.4) return PvEResultado.TENTOU_FALHOU;
  return PvEResultado.ERRADO;
}

function _sugerirProxima(resultado, foco) {
  const alvo = foco.parcialNome || foco.parcialId;
  switch (resultado) {
    case PvEResultado.CONSEGUIU:     return `manter ${alvo} — ação consolidada`;
    case PvEResultado.PARCIAL:       return `repetir ação em ${alvo} com mais consistência`;
    case PvEResultado.TENTOU_FALHOU: return `executou a ideia mas errou dose — ajustar intensidade em ${alvo}`;
    case PvEResultado.ERRADO:        return `reavaliar ação planejada em ${alvo}`;
    case PvEResultado.SEM_DADOS:     return `coletar mais voltas válidas em ${alvo}`;
    default:                         return null;
  }
}

File: ios/p1fast-core/Sources/P1FastCore/PlannedVsExecuted.swift
// ═══════════════════════════════════════════════════════════
// PlannedVsExecuted — port Swift de
//   src/domain/planned-vs-executed.js
// ═══════════════════════════════════════════════════════════
// Avaliação pedagógica do plano: para cada foco do plano (1 por
// PARCIAL), analisa as execuções reais e emite veredicto:
//   CONSEGUIU / PARCIAL / TENTOU_FALHOU / ERRADO / SEM_DADOS
//
// MS-6.4 (port 2026-05-04). Camada pura — caller traz plan + laps
// + benchmark.

import Foundation

public enum PvEResultado: String, Sendable, Codable, CaseIterable {
    case conseguiu     = "conseguiu"
    case parcial       = "parcial"
    case tentouFalhou  = "tentou-falhou"
    case errado        = "errado"
    case semDados      = "sem-dados"
}

public struct PlanFoco: Sendable, Equatable {
    public let parcialId: String
    public let parcialNome: String
    public let acaoPrincipal: String?
    public let justificativa: String?

    public init(
        parcialId: String, parcialNome: String,
        acaoPrincipal: String? = nil, justificativa: String? = nil
    ) {
        self.parcialId = parcialId
        self.parcialNome = parcialNome
        self.acaoPrincipal = acaoPrincipal
        self.justificativa = justificativa
    }
}

public struct PedagogicalPlan: Sendable, Equatable {
    public let metaPrincipal: String?
    public let criterioSucesso: String?
    public let focos: [PlanFoco]

    public init(
        metaPrincipal: String? = nil,
        criterioSucesso: String? = nil,
        focos: [PlanFoco]
    ) {
        self.metaPrincipal = metaPrincipal
        self.criterioSucesso = criterioSucesso
        self.focos = focos
    }
}

public struct PvELap: Sendable, Equatable {
    public let id: String
    public let numero: Int?
    public let valida: Bool
    public let temposPorParcial: [String: Double?]

    public init(id: String, numero: Int? = nil, valida: Bool, temposPorParcial: [String: Double?] = [:]) {
        self.id = id
        self.numero = numero
        self.valida = valida
        self.temposPorParcial = temposPorParcial
    }
}

public struct PvEExecucao: Sendable, Equatable {
    public let lapId: String
    public let numero: Int?
    public let tempoMs: Double?
    public let classificacaoPrincipal: ErroTipo?
}

public struct PvEFocoResult: Sendable, Equatable {
    public let parcialId: String
    public let parcialNome: String
    public let acaoAlvo: String?
    public let justificativa: String?
    public let execucoes: [PvEExecucao]
    public let resultado: PvEResultado
    public let proximaAcao: String?
}

public struct PvEResult: Sendable, Equatable {
    public let metaPrincipal: String?
    public let criterioSucesso: String?
    public let focos: [PvEFocoResult]
}

private struct AcaoMapping {
    let alinhadas: [ErroTipo]
    let tentativas: [ErroTipo]
}

public enum PlannedVsExecuted {

    /// Avalia o plano contra as voltas + benchmark. `nil` se plan inválido.
    public static func evaluate(
        plan: PedagogicalPlan?,
        laps: [PvELap],
        benchmark: BenchmarkResult?
    ) -> PvEResult? {
        guard let plan else { return nil }
        let validLaps = laps.filter(\.valida)
        var focos: [PvEFocoResult] = []
        for foco in plan.focos {
            let mapping = mapAcao(foco.acaoPrincipal)
            let execucoes: [PvEExecucao] = validLaps.map { l in
                let tempoExec = l.temposPorParcial[foco.parcialId] ?? nil
                let tempoRef = benchmark?.melhorPorParcial[foco.parcialId]??.tempoMs
                let principal: ErroTipo? = {
                    guard let ref = tempoRef, ref > 0,
                          let exec = tempoExec, exec > 0 else { return nil }
                    let dTempo = (exec - ref) / ref
                    if dTempo <= 0.01      { return .manteveLinha }
                    else if dTempo <= 0.03 { return .boaFrenagem }
                    else                   { return nil }
                }()
                return PvEExecucao(
                    lapId: l.id, numero: l.numero,
                    tempoMs: tempoExec, classificacaoPrincipal: principal
                )
            }
            let classificados = execucoes.compactMap { ex -> ErroTipo? in
                ex.classificacaoPrincipal
            }
            let resultado = julgar(mapping: mapping, classificados: classificados)
            let proxima = sugerirProxima(resultado: resultado, foco: foco)
            focos.append(PvEFocoResult(
                parcialId: foco.parcialId,
                parcialNome: foco.parcialNome,
                acaoAlvo: foco.acaoPrincipal,
                justificativa: foco.justificativa,
                execucoes: execucoes,
                resultado: resultado,
                proximaAcao: proxima
            ))
        }
        return PvEResult(
            metaPrincipal: plan.metaPrincipal,
            criterioSucesso: plan.criterioSucesso,
            focos: focos
        )
    }
}

// MARK: - Mapping ações → ErroTipo

private func acaoMap() -> [(key: String, mapping: AcaoMapping)] {
    [
        ("atrasar frenagem",
            AcaoMapping(alinhadas: [.boaFrenagem, .manteveLinha], tentativas: [.freouTardeDemais])),
        ("antecipar frenagem",
            AcaoMapping(alinhadas: [.manteveLinha, .boaFrenagem], tentativas: [.freouCedo])),
        ("acelerar mais cedo",
            AcaoMapping(alinhadas: [.saiuForte, .manteveLinha], tentativas: [.perdeuTracao])),
        ("saída mais forte",
            AcaoMapping(alinhadas: [.saiuForte, .manteveLinha], tentativas: [.perdeuTracao])),
        ("entrar mais forte",
            AcaoMapping(alinhadas: [.boaFrenagem, .manteveLinha], tentativas: [.entrouForteDemais])),
        ("manter linha",
            AcaoMapping(alinhadas: [.manteveLinha], tentativas: [])),
        ("reduzir tempo",
            AcaoMapping(alinhadas: [.manteveLinha, .boaFrenagem, .saiuForte], tentativas: [])),
        ("trabalhar execução",
            AcaoMapping(alinhadas: [.manteveLinha, .boaFrenagem, .saiuForte], tentativas: [])),
        ("focar em",
            AcaoMapping(alinhadas: [.manteveLinha, .boaFrenagem, .saiuForte], tentativas: [])),
    ]
}

private func mapAcao(_ acao: String?) -> AcaoMapping? {
    guard let raw = acao?.lowercased() else { return nil }
    for (key, mapping) in acaoMap() where raw.contains(key) {
        return mapping
    }
    return nil
}

private func julgar(mapping: AcaoMapping?, classificados: [ErroTipo]) -> PvEResultado {
    if classificados.isEmpty { return .semDados }
    guard let mapping else {
        let ok = classificados.filter { $0 == .manteveLinha }.count
        let pct = Double(ok) / Double(classificados.count)
        if pct >= 0.6 { return .parcial }
        return .semDados
    }
    let n = Double(classificados.count)
    let alinhadas = classificados.filter { mapping.alinhadas.contains($0) }.count
    let tentativas = classificados.filter { mapping.tentativas.contains($0) }.count
    let pctAlinhadas = Double(alinhadas) / n
    let pctTentativas = Double(tentativas) / n
    if pctAlinhadas >= 0.6 { return .conseguiu }
    if pctAlinhadas >= 0.3 { return .parcial }
    if pctTentativas >= 0.4 { return .tentouFalhou }
    return .errado
}

private func sugerirProxima(resultado: PvEResultado, foco: PlanFoco) -> String {
    let alvo = foco.parcialNome.isEmpty ? foco.parcialId : foco.parcialNome
    switch resultado {
    case .conseguiu:     return "manter \(alvo) — ação consolidada"
    case .parcial:       return "repetir ação em \(alvo) com mais consistência"
    case .tentouFalhou:  return "executou a ideia mas errou dose — ajustar intensidade em \(alvo)"
    case .errado:        return "reavaliar ação planejada em \(alvo)"
    case .semDados:      return "coletar mais voltas válidas em \(alvo)"
    }
}
